Add support for GCP authentication for GKE

# Summary

Add support for GCP authentication for GKE instead of keeping a token in a secret, pretty much similar to the AWS EKS support.

# Motivation

Some GKE scenarios with multi cluster management can rely on Anthos Connect Gateway like seen here: https://cloud.google.com/anthos/multicluster-management/gateway/using. In such scenarios, the only authentication option is via GCP authentication, sending an OAuth token instead of a Kubernetes token. 

This will open endless possibilities like having Argo CD in GCP orchestrating the deployment of many environments in multiple places (on-prem, GCP, other clouds, etc) via Anthos Connect Gateway.

# Proposal

Leverage GCP authentication similar to what is done in aws eks get-token nowadays. I believe the K8S client-go already supports the auth plugin for GCP.

@laloyalo 

I've implemented what I believe is needed, however when I try to run the make test, I'm getting the following timeout. I've ran multiple times, same result and if I try to do this git command by hand it works. Any idea?

--- FAIL: TestVerifyCommitSignature (92.16s)
    git_test.go:287: 
                Error Trace:    git_test.go:287
                Error:          Received unexpected error:
                                `git fetch origin --tags --force` failed timeout after 1m30s
                Test:           TestVerifyCommitSignature
    git_test.go:293: 
                Error Trace:    git_test.go:293
                Error:          Received unexpected error:
                                `git checkout --force ae2d0ff0a6ac34dc3e8493b2ad45a9badc61a26c` failed exit status 128: fatal: reference is not a tree: ae2d0ff0a6ac34dc3e8493b2ad45a9badc61a26c
                Test:           TestVerifyCommitSignature
    git_test.go:301: 
                Error Trace:    git_test.go:301
                Error:          "error: 28027897aad1262662096745f2ce2d4c74d02b7f: unable to read file." does not contain "gpg: Signature made"
                Test:           TestVerifyCommitSignature
    git_test.go:308: 
                Error Trace:    git_test.go:308
                Error:          Should be empty, but was error: 85d660f0b967960becce3d49bd51c678ba2a5d24: unable to read file.
                Test:           TestVerifyCommitSignature
FAIL

--- a/cmd/argocd-k8s-auth/commands/argocd_k8s_auth.go
+++ b/cmd/argocd-k8s-auth/commands/argocd_k8s_auth.go
@@ -19,6 +19,7 @@ func NewCommand() *cobra.Command {
 	}
 
 	command.AddCommand(newAWSCommand())
+	command.AddCommand(newGCPCommand())
 
 	return command
 }

--- a/cmd/argocd-k8s-auth/commands/gcp.go
+++ b/cmd/argocd-k8s-auth/commands/gcp.go
@@ -1,0 +1,40 @@
+package commands
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+	"golang.org/x/oauth2/google"
+
+	"github.com/argoproj/argo-cd/v2/util/errors"
+)
+
+var (
+	// defaultGCPScopes:
+	// - cloud-platform is the base scope to authenticate to GCP.
+	// - userinfo.email is used to authenticate to GKE APIs with gserviceaccount
+	//   email instead of numeric uniqueID.
+	// https://github.com/kubernetes/client-go/blob/be758edd136e61a1bffadf1c0235fceb8aee8e9e/plugin/pkg/client/auth/gcp/gcp.go#L59
+	defaultGCPScopes = []string{
+		"https://www.googleapis.com/auth/cloud-platform",
+		"https://www.googleapis.com/auth/userinfo.email",
+	}
+)
+
+func newGCPCommand() *cobra.Command {
+	var command = &cobra.Command{
+		Use: "gcp",
+		Run: func(c *cobra.Command, args []string) {
+			// Preferred way to retrieve GCP credentials
+			// https://github.com/golang/oauth2/blob/9780585627b5122c8cc9c6a378ac9861507e7551/google/doc.go#L54-L68
+			cred, err := google.FindDefaultCredentials(context.Background(), defaultGCPScopes...)
+			errors.CheckError(err)
+			token, err := cred.TokenSource.Token()
+			errors.CheckError(err)
+			_, _ = fmt.Fprint(os.Stdout, formatJSON(token.AccessToken, token.Expiry))
+		},
+	}
+	return command
+}

--- a/docs/operator-manual/declarative-setup.md
+++ b/docs/operator-manual/declarative-setup.md
@@ -554,6 +554,35 @@ stringData:
     }
 ```
 
+GKE cluster secret example using argocd-k8s-auth and [Workload Identity](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity):
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: mycluster-secret
+  labels:
+    argocd.argoproj.io/secret-type: cluster
+type: Opaque
+stringData:
+  name: mycluster.com
+  server: https://mycluster.com
+  config: |
+    {
+      "execProviderConfig": {
+        "command": "argocd-k8s-auth",
+        "args": ["gcp"],
+        "apiVersion": "client.authentication.k8s.io/v1beta1"
+      },
+      "tlsClientConfig": {
+        "insecure": false,
+        "caData": "<base64 encoded certificate>"
+      }
+    }
+```
+
+Note that you must enable Workload Identity on your GKE cluster, create GCP service account with appropriate IAM role and bind it to Kubernetes service account for argocd-application-controller and argocd-server (showing Pod logs on UI). See [Use Workload Identity](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity) and [Authenticating to the Kubernetes API server](https://cloud.google.com/kubernetes-engine/docs/how-to/api-server-authentication).
+
 ## Helm Chart Repositories
 
 Non standard Helm Chart repositories have to be registered explicitly.

